### PR TITLE
Support OR-ed condition in Delta checkpoint iterator

### PR DESCRIPTION
**Pull request by findinpath**
## Description

Relates to https://github.com/trinodb/trino/issues/19156

Very much inspired from https://github.com/trinodb/trino/pull/19240

The problem this PR is solving is that it avoids working with `TupleDomain.all()` as tuple domain when there are more predicates specified for the `CheckpointEntryIterator`:

https://github.com/trinodb/trino/blob/d23921d86777fbf778e7c3f63edcbd293c501971/plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java#L186-L187

What this means is that in such cases, all the entries from the Delta Lake Parquet checkpoint file will be returned as results even though the code flow is interested in a small subset of the entries.

A common use-case  of specifying more predicates (entry types) for the `CheckpointEntryIterator` is `DeltaLakeMetadata.getTableHandle()`:

https://github.com/trinodb/trino/blob/d23921d86777fbf778e7c3f63edcbd293c501971/plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java#L495-L498

## Additional context and related issues

The `ParquetPageSourceFactory` has been slightly refactored to expose a method handling multiple tuple domains.
The results returned in the page source are corresponding to any of the specified tuple domains.

To put things in perspective, the changes from this PR have been tested against a 300MB checkpoint file  to see how it performs.
The example showcased below depends on a local file, but the changes have been tested as well against a S3 bucket and for retrieving `metadata` & `protocol` together there is no relevant performance gain.

tldr; performance wise the changes coming with this PR are rather modest because the amount of bytes read from the Parquet columnar file for `metadata`  and `protocol` is pretty small compared to the bytes needed to encode the `add` entries.

```
    @Test
    public void testBytesRead()
            throws IOException
    {
        String targetPath = "file:///home/marius/Downloads/00000000000000000003.checkpoint.parquet";

        CheckpointEntryIterator metadataEntryIterator =
                createCheckpointEntryIterator(URI.create(targetPath), ImmutableSet.of(METADATA), Optional.empty(), Optional.empty());
        CheckpointEntryIterator protocolEntryIterator =
                createCheckpointEntryIterator(URI.create(targetPath), ImmutableSet.of(PROTOCOL), Optional.empty(), Optional.empty());
        MetadataEntry metadataEntry = getOnlyElement(metadataEntryIterator).getMetaData();
        ProtocolEntry protocolEntry = getOnlyElement(protocolEntryIterator).getProtocol();

        CheckpointEntryIterator metadataAndProtocolEntryIterator =
                createCheckpointEntryIterator(URI.create(targetPath), ImmutableSet.of(METADATA, PROTOCOL), Optional.empty(), Optional.empty());
        CheckpointEntryIterator addEntryIterator = createCheckpointEntryIterator(
                URI.create(targetPath),
                ImmutableSet.of(ADD),
                Optional.of(metadataEntry),
                Optional.of(protocolEntry));

        System.out.println("METADATA");
        System.out.println("completed positions: " + metadataEntryIterator.getCompletedPositions());
        System.out.println("completed bytes: " + metadataEntryIterator.getCompletedBytes());
        System.out.println("PROTOCOL");
        System.out.println("completed positions: " + protocolEntryIterator.getCompletedPositions());
        System.out.println("completed bytes: " + protocolEntryIterator.getCompletedBytes());
        System.out.println("METADATA&PROTOCOL");
        long startMetadataAndProtocol = System.currentTimeMillis();
        System.out.println("size: " + Iterators.size(metadataAndProtocolEntryIterator));
        long endMetadataAndProtocol = System.currentTimeMillis();
        System.out.println("Elapsed Time in milli seconds: " + (endMetadataAndProtocol - startMetadataAndProtocol));
        System.out.println("completed positions: " + metadataAndProtocolEntryIterator.getCompletedPositions());
        System.out.println("completed bytes: " + metadataAndProtocolEntryIterator.getCompletedBytes());
        System.out.println("ADD");
        long startAdd = System.currentTimeMillis();
        System.out.println("size: " + Iterators.size(addEntryIterator));
        long endAdd = System.currentTimeMillis();
        System.out.println("Elapsed Time in milli seconds: " + (endAdd - startAdd));
        System.out.println("completed positions: " + addEntryIterator.getCompletedPositions());
        System.out.println("completed bytes: " + addEntryIterator.getCompletedBytes());
    }
```

The results on testing with `master`
```
METADATA
completed positions: OptionalLong[470000]
completed bytes: 49876
PROTOCOL
completed positions: OptionalLong[470000]
completed bytes: 49224
METADATA&PROTOCOL
size: 2
Elapsed Time in milli seconds: 261
completed positions: OptionalLong[1235156]
completed bytes: 50782
ADD
size: 1235154
Elapsed Time in milli seconds: 14040
completed positions: OptionalLong[1235156]
completed bytes: 329586188
```

The results on testing with the changes from this PR:
```
METADATA
completed positions: OptionalLong[470000]
completed bytes: 49876
PROTOCOL
completed positions: OptionalLong[470000]
completed bytes: 49224
METADATA&PROTOCOL
size: 2
Elapsed Time in milli seconds: 121
completed positions: OptionalLong[470000]
completed bytes: 49948
ADD
size: 1235154
Elapsed Time in milli seconds: 14194
completed positions: OptionalLong[1235156]
completed bytes: 329586188
```

I slightly modified the example above to test against MinIO and check what kind of conversations are made with the object storage:

```
        MinioClient minioClient = new MinioClient("http://localhost:9080", "minio-access-key", "minio-secret-key");
        minioClient.captureBucketNotifications(BUCKET_NAME, event -> {
            System.out.println(event.eventType() + " - " + event.requestParameters());
        });
```

Relevant output while running with this PR

```
METADATA
s3:ObjectAccessed:Head - {principalId=minio-access-key, region=, sourceIPAddress=172.28.0.1}
s3:ObjectAccessed:Get - {principalId=minio-access-key, range=bytes=-49152, region=, sourceIPAddress=172.28.0.1}
s3:ObjectAccessed:Get - {principalId=minio-access-key, range=bytes=4-727, region=, sourceIPAddress=172.28.0.1}
completed positions: OptionalLong[470000]
completed bytes: 49876
PROTOCOL
s3:ObjectAccessed:Head - {principalId=minio-access-key, region=, sourceIPAddress=172.28.0.1}
s3:ObjectAccessed:Get - {principalId=minio-access-key, range=bytes=-49152, region=, sourceIPAddress=172.28.0.1}
s3:ObjectAccessed:Get - {principalId=minio-access-key, range=bytes=728-799, region=, sourceIPAddress=172.28.0.1}
completed positions: OptionalLong[470000]
completed bytes: 49224
METADATA&PROTOCOL
s3:ObjectAccessed:Head - {principalId=minio-access-key, region=, sourceIPAddress=172.28.0.1}
s3:ObjectAccessed:Get - {principalId=minio-access-key, range=bytes=-49152, region=, sourceIPAddress=172.28.0.1}
s3:ObjectAccessed:Get - {principalId=minio-access-key, range=bytes=4-799, region=, sourceIPAddress=172.28.0.1}
size: 2
Elapsed Time in milli seconds: 126
completed positions: OptionalLong[470000]
completed bytes: 49948
```

Relevant output while running on `master`

```
METADATA
s3:ObjectAccessed:Head - {principalId=minio-access-key, region=, sourceIPAddress=172.28.0.1}
s3:ObjectAccessed:Get - {principalId=minio-access-key, range=bytes=-49152, region=, sourceIPAddress=172.28.0.1}
s3:ObjectAccessed:Get - {principalId=minio-access-key, range=bytes=4-727, region=, sourceIPAddress=172.28.0.1}
completed positions: OptionalLong[470000]
completed bytes: 49876
PROTOCOL
s3:ObjectAccessed:Head - {principalId=minio-access-key, region=, sourceIPAddress=172.28.0.1}
s3:ObjectAccessed:Get - {principalId=minio-access-key, range=bytes=-49152, region=, sourceIPAddress=172.28.0.1}
s3:ObjectAccessed:Get - {principalId=minio-access-key, range=bytes=728-799, region=, sourceIPAddress=172.28.0.1}
completed positions: OptionalLong[470000]
completed bytes: 49224
METADATA&PROTOCOL
s3:ObjectAccessed:Head - {principalId=minio-access-key, region=, sourceIPAddress=172.28.0.1}
s3:ObjectAccessed:Get - {principalId=minio-access-key, range=bytes=-49152, region=, sourceIPAddress=172.28.0.1}
s3:ObjectAccessed:Get - {principalId=minio-access-key, range=bytes=4-799, region=, sourceIPAddress=172.28.0.1}
s3:ObjectAccessed:Get - {principalId=minio-access-key, range=bytes=125355536-125355952, region=, sourceIPAddress=172.28.0.1}
s3:ObjectAccessed:Get - {principalId=minio-access-key, range=bytes=250711810-250712226, region=, sourceIPAddress=172.28.0.1}
size: 2
Elapsed Time in milli seconds: 355
completed positions: OptionalLong[1235156]
completed bytes: 50782
```

For `METADATA` & `PROTOCOL` when specified together, for the file I am using for testing, the `CheckpointEntryIterator makes currently (in `master`) the extra call:
```
s3:ObjectAccessed:Get - {principalId=minio-access-key, range=bytes=125355536-125355952, region=, sourceIPAddress=172.28.0.1}
```

## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
#
* ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java [hunk 1]
```diff
@@ -238,37 +274,60 @@ public static ReaderPageSource createPageSource(
...
+                                    columnIndex,
+                                    bloomFilterStore,
+                                    timeZone,
+                                    domainCompactionThreshold)) {
+                        blocks.add(block);
+                        blockStarts.add(nextStart);
+                        columnIndexes.add(columnIndex);
+                        isBlockAlreadyAdded = true;
+                    }
+                    nextStart += block.getRowCount();
```
Comment (jkylling) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:329]: Is it correct that we increment this for every `parquetDomain`?

File: lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java [hunk 2]
```diff
@@ -671,7 +671,7 @@ private static Optional<Collection<Object>> extractDiscreteValues(int domainComp
         return Optional.of(valueSet.getDiscreteSet());
     }
 
-    private FilterPredicate convertToParquetFilter(DateTimeZone timeZone)
+    public FilterPredicate convertToParquetFilter(DateTimeZone timeZone)
```
Comment (findepi) [inline lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:674]: Undo

This method is not good to expose, since it returns nullable.

use `toParquetFilter` defined above which delegates here and wraps with optional

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java [hunk 3]
```diff
@@ -219,6 +221,40 @@ public static ReaderPageSource createPageSource(
...
+                domainCompactionThreshold,
+                estimatedFileSize);
+    }
+
+    /**
+     * This method is available for other callers to use directly.
+     * The results in the resulting page source will match any of the
+     * predicates specified in the list `effectivePredicates`.
+     */
+    public static ReaderPageSource createPageSource(
```
Comment (jkylling) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:245]: It feels like this should be in a separate commit with some tests with effectivePredicates.size > 1.

**Review by findepi — COMMENTED** [review]
"Accept FilterPredicate in ParquetReader"

**Review by jkylling — COMMENTED** [review]
Thank you for fixing this!

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java [hunk 4]
```diff
@@ -238,37 +274,60 @@ public static ReaderPageSource createPageSource(
...
+
+                FilterPredicate filter = parquetPredicate.convertToParquetFilter(timeZone);
+                if (filter != null) {
+                    if (nullableFilterPredicate == null) {
+                        nullableFilterPredicate = filter;
+                    }
+                    else {
+                        nullableFilterPredicate = FilterApi.or(nullableFilterPredicate, filter);
+                    }
+                }
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:294]: Move this after the loop, then it becomes simpler

```java
Optional<FilterPredicate> filterPredicate = parquetDomains.stream()
                    .map(ParquetDomainAndPredicate::predicate)
                    .map(parquetPredicate -> parquetPredicate.toParquetFilter(timeZone))
                    .flatMap(Optional::stream)
                    // TODO create balanced tree when this is used with high number of filters. Currently it's used with at most 2
                    .reduce(FilterApi::or);
```

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java [hunk 5]
```diff
@@ -238,37 +274,60 @@ public static ReaderPageSource createPageSource(
...
-                    blockStarts.add(nextStart);
-                    columnIndexes.add(columnIndex);
+                boolean isBlockAlreadyAdded = false;
+                for (ParquetDomainAndPredicate domain : parquetDomains) {
+                    if (isBlockAlreadyAdded) {
+                        break;
+                    }
+
+                    TupleDomain<ColumnDescriptor> parquetTupleDomain = domain.tupleDomain;
+                    TupleDomainParquetPredicate parquetPredicate = domain.predicate;
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:310]: indentiation below is off. make sure to format the file

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java [hunk 6]
```diff
@@ -238,37 +274,60 @@ public static ReaderPageSource createPageSource(
...
 
             Map<List<String>, ColumnDescriptor> descriptorsByPath = getDescriptors(fileSchema, requestedSchema);
-            TupleDomain<ColumnDescriptor> parquetTupleDomain = options.isIgnoreStatistics()
-                    ? TupleDomain.all()
-                    : getParquetTupleDomain(descriptorsByPath, effectivePredicate, fileSchema, useColumnNames);
-
-            TupleDomainParquetPredicate parquetPredicate = buildPredicate(requestedSchema, parquetTupleDomain, descriptorsByPath, timeZone);
+            List<ParquetDomainAndPredicate> parquetDomains = new ArrayList<>();
+            FilterPredicate nullableFilterPredicate = null;
+            for (TupleDomain<HiveColumnHandle> domain : effectivePredicates) {
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:279]: `options.isIgnoreStatistics()` can be consulted once.
it never changes. would make the code simpler to follow.

Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:279]:
```java
List<ParquetDomainAndPredicate> domainAndPredicates;
            Optional<FilterPredicate> filterPredicate;
            if (!options.isIgnoreStatistics()) {
                domainAndPredicates = effectivePredicates.stream()
                        .map(domain -> {
                            TupleDomain<ColumnDescriptor> parquetTupleDomain = getParquetTupleDomain(descriptorsByPath, domain, fileSchema, useColumnNames);
                            TupleDomainParquetPredicate parquetPredicate = buildPredicate(requestedSchema, parquetTupleDomain, descriptorsByPath, timeZone);
                            return new ParquetDomainAndPredicate(parquetTupleDomain, parquetPredicate);
                        })
                        .collect(toImmutableList());

                filterPredicate = domainAndPredicates.stream()
                        .map(ParquetDomainAndPredicate::predicate)
                        .map(parquetPredicate -> parquetPredicate.toParquetFilter(timeZone))
                        .flatMap(Optional::stream)
                        // TODO create balanced tree when this is used with high number of filters. Currently it's used with at most 2
                        .reduce(FilterApi::or);
            }
            else {
                domainAndPredicates = ImmutableList.of();
                filterPredicate = Optional.empty();
            }
```

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java [hunk 7]
```diff
@@ -238,37 +274,60 @@ public static ReaderPageSource createPageSource(
...
 
             long nextStart = 0;
             ImmutableList.Builder<BlockMetaData> blocks = ImmutableList.builder();
             ImmutableList.Builder<Long> blockStarts = ImmutableList.builder();
             ImmutableList.Builder<Optional<ColumnIndexStore>> columnIndexes = ImmutableList.builder();
             for (BlockMetaData block : parquetMetadata.getBlocks()) {
                 long firstDataPage = block.getColumns().get(0).getFirstDataPageOffset();
-                Optional<ColumnIndexStore> columnIndex = getColumnIndexStore(dataSource, block, descriptorsByPath, parquetTupleDomain, options);
-                Optional<BloomFilterStore> bloomFilterStore = getBloomFilterStore(dataSource, block, parquetTupleDomain, options);
-
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:254]: undo

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeTransactionLogEntry.java [hunk 8]
```diff
@@ -156,6 +158,25 @@ public DeltaLakeTransactionLogEntry withCommitInfo(CommitInfoEntry commitInfo)
         return new DeltaLakeTransactionLogEntry(txn, add, remove, metaData, protocol, commitInfo, cdcEntry);
     }
 
+    @Override
+    public boolean equals(Object o)
```
[the remark this answers is not in the document]
Comment (findinpath) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeTransactionLogEntry.java:162]: `io.trino.plugin.deltalake.transactionlog.checkpoint.TestCheckpointEntryIterator#testReadMetadataAndProtocolEntry`

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java (hunk 1 above)
Comment (findinpath) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:329]: Good catch.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java (hunk 7 above)
Comment (findinpath) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:254]: I don't follow the suggestion

**Comment by findepi** [comment]
> master
METADATA&PROTOCOL
completed bytes: 50782


> changes from this PR:
METADATA&PROTOCOL
completed bytes: 49948

in both cases we read ~50k of 300MB file

**Comment by jkylling** [comment]
> > master
> > METADATA&PROTOCOL
> > completed bytes: 50782
> 
> > changes from this PR:
> > METADATA&PROTOCOL
> > completed bytes: 49948
> 
> in both cases we read ~50k of 300MB file

The number of bytes read should be approximately the same, but the number of completed positions is very different (1235156 vs. 470000). The bottle neck is that we are iterating over all the rows of the row groups we read, and checking if they contain non-null data for the metadata or protocol entries, but this is only the case for two rows. On master, since the columns we select are all null except in two rows, it's cheap in terms of data, but it's expensive in terms of compute (relative to what this operation is supposed to do), as we iterate over 1235156 rows.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java [hunk 9]
```diff
@@ -237,35 +239,56 @@ public static ReaderPageSource createPageSource(
...
+                        .map(ParquetDomainAndPredicate::predicate)
+                        .map(parquetPredicate -> parquetPredicate.toParquetFilter(timeZone))
+                        .flatMap(Optional::stream)
+                        // TODO create balanced tree when this is used with high number of filters.
+                        .reduce(FilterApi::or);
+            }
+            else {
+                domainAndPredicates = ImmutableList.of();
+                filterPredicate = Optional.empty();
+            }
```
Comment (raunaqmorarka) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:263]: We can initialise to these values and avoid the `else`

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java [hunk 10]
```diff
@@ -237,35 +239,56 @@ public static ReaderPageSource createPageSource(
...
-                        columnIndex,
-                        bloomFilterStore,
-                        timeZone,
-                        domainCompactionThreshold)) {
-                    blocks.add(block);
-                    blockStarts.add(nextStart);
-                    columnIndexes.add(columnIndex);
+                for (ParquetDomainAndPredicate domain : domainAndPredicates) {
+                    TupleDomain<ColumnDescriptor> parquetTupleDomain = domain.tupleDomain;
+                    TupleDomainParquetPredicate parquetPredicate = domain.predicate;
```
Comment (raunaqmorarka) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:273]: Just use the record accessors directly wherever needed instead of extracting a variable

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java (hunk 9 above)
Comment (raunaqmorarka) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:263]: Or at least invert the `if` to avoid the negation

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java [hunk 11]
```diff
@@ -237,35 +239,56 @@ public static ReaderPageSource createPageSource(
...
+                            return new ParquetDomainAndPredicate(parquetTupleDomain, parquetPredicate);
+                        })
+                        .collect(toImmutableList());
+
+                filterPredicate = domainAndPredicates.stream()
+                        .map(ParquetDomainAndPredicate::predicate)
+                        .map(parquetPredicate -> parquetPredicate.toParquetFilter(timeZone))
+                        .flatMap(Optional::stream)
+                        // TODO create balanced tree when this is used with high number of filters.
+                        .reduce(FilterApi::or);
```
Comment (raunaqmorarka) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:258]: We're not actually using `filterPredicate` in this class, so why not let ParquetReader handle creating this internally ?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java [hunk 12]
```diff
@@ -179,20 +178,20 @@ public CheckpointEntryIterator(
...
 
-        List<HiveColumnHandle> columns = fields.stream()
-                .map(field -> buildColumnHandle(field, checkpointSchemaManager, this.metadataEntry, this.protocolEntry).toHiveColumnHandle())
-                .collect(toImmutableList());
-
-        TupleDomain<HiveColumnHandle> tupleDomain = columns.size() > 1 ?
-                TupleDomain.all() :
-                buildTupleDomainColumnHandle(getOnlyElement(fields), getOnlyElement(columns));
+        ImmutableList.Builder<HiveColumnHandle> columnsBuilder = ImmutableList.builderWithExpectedSize(fields.size());
+        ImmutableList.Builder<TupleDomain<HiveColumnHandle>> tupleDomainBuilder = ImmutableList.builderWithExpectedSize(fields.size());
```
Comment (raunaqmorarka) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java:182]: tupleDomainBuilder -> disjunctDomainsBuilder

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeTransactionLogEntry.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeTransactionLogEntry.java:162]: Maybe add this in a prep commit ?

**Review by raunaqmorarka — APPROVED** [review]

**Comment by findepi** [comment]
thanks @jkylling for the feedback and measuring the times

> in both cases we read ~50k of 300MB file

spoke with @ebyhr and realized we measured only the amount of data loaded, but not the number of round trips to the storage. 
It the small difference in number of bytes read is effect of saving a bunch of small s3 requests, then this is still an important improvement

@findinpath please measure that (eg with MinIO notifications / request log)

As @findinpath spoke earlier, Open question remains: could we achieve same improvement within ParquetReader, based on column select and all-null information for row groups. This would mean same improvement for chckpoint reading without any checkpoint reader changes, and also for other cases.

**Comment by findinpath** [comment]
> @findinpath please measure that (eg with MinIO notifications / request log)

@findepi  updated the description 


> For `METADATA` & `PROTOCOL` when specified together, for the file I am using for testing, the `CheckpointEntryIterator makes currently (in `master`) the extra call:
```
s3:ObjectAccessed:Get - {principalId=minio-access-key, range=bytes=125355536-125355952, region=, sourceIPAddress=172.28.0.1}
```

**Comment by findinpath** [comment]
> Open question remains: could we achieve same improvement within ParquetReader, based on column select and all-null information for row groups. 

We do already have the null/empty check here:

https://github.com/trinodb/trino/blob/f96528757114e08e1d83ca18e52b79424c45a827/lib/trino-parquet/src/main/java/io/trino/parquet/predicate/PredicateUtils.java#L143-L147

**Comment by findinpath** [comment]
Rebasing on `master` to resolve conflicts on `CheckpointEntryIterator`

**Comment by findepi** [comment]
per offline discussion with @raunaqmorarka 
the alternative -- to make ParquetReader skip row groups when stats indicate column is all null -- is considered complicated, so we should probably land this
